### ISSUE TITLE
Potential OAuth2 issue, allow/deny, and callback

### ISSUE THREAD

**Issue by joho1968** [mattermost/mattermost#16646]
#### Summary
Odd OAuth2 behavior when user selects "Deny" on authorization prompt

#### Steps to reproduce
Mattermost Server 5.30.1

1. Use an OAuth2 client (in this case, I'm writing my own)
2. Select to authenticate via Mattermost
3. Mattermost login is displayed
4. Entere credentials
5. Mattermost displays prompt, Allow/Deny
6. Select Deny
7. Control is returned to OAuth2 client
8. Repeat; this time Mattermost re-directs back to client without error message

#### Expected behavior
I would have expected Mattermost to prompt again, since I've previously selected "Deny".

Or, at the very least, return to client with error message "access_denied".

#### Observed behavior (that appears unintentional) 
Mattermost immediately issues callback (back to client) without error message.

#### Possible fixes
Return to client with error message indicating "access_denied", alternatively, present prompt again.

I initially posted this as a question om the developers channel:
https://community-daily.mattermost.com/core/pl/rzm3ifwzjjgaik7afa14sgpkbw

**Comment by ThiefMaster** [comment]
Remembering a "deny" choice in an oauth consent prompt seems like a very bad idea.

Those prompts are pretty much *always* triggered by a user action - clicking "login with mattermost", accessing a protected site that immediately redirects to oauth login, etc. - and regardless of the reason you denied the request (misclicked, decided to not login with mattermost at that time), there's simply no reason to remember this. When you come back there you can just deny again - there is absolutely no advantage in automatically failing and redirecting back to the oauth client app.

In fact, I would *not* expect an oauth server to store any data related to my and that particular oauth app if I did not granted the app access to my account...

**Comment by joho1968** [comment]
Fair enough. But I would expect, at the very least, that Mattermost instead present the prompt again then, which it does not do. I suspect the reason it does not do that is because I'm effectively already logged in to Mattermost since the prompt appears after the login. So regardless of how Mattermost handles this going forward, it's quite broken now (IMHO) since it puts the prompt out of order if you select Deny.

**Comment by ThiefMaster** [comment]
Yes, that's what I meant - IMHO "deny" should not store anything, so when you try again you get the prompt again.

**Comment by amyblais** [comment]
Ticket for reference https://mattermost.atlassian.net/browse/MM-42501.

**Comment by mkraft** [comment]
@joho1968 Would you be up for submitting a fix for this?

**Comment by joho1968** [comment]
Wow ... a blast from the past 😊

Unfortunately, I don't "Go"; I do a lot of other languages, but not Golang.

But surely, this fix can't be that complex, since it's about **not** remembering the choice?

**Comment by hanzei** [comment]
After multiple years of inactivity, we're closing this to help keep our issue tracker focused on active work.
